Source/Core/VideoBackends/Metal/MTLUtil.mm

```diff
@@ -0,0 +1,470 @@
+  resource.msl_sampler = msl_sampler;
+  return resource;
+}
+
+std::string Metal::Util::TranslateShaderToMSL(ShaderStage stage, std::string_view source)
+{
+  std::string full_source;
+
+  std::string_view header = stage == ShaderStage::Compute ? COMPUTE_SHADER_HEADER : SHADER_HEADER;
+  full_source.reserve(header.size() + SUBGROUP_HELPER_HEADER.size() + source.size());
+
+  full_source.append(header);
+  if (Metal::g_features.subgroup_ops)
+    full_source.append(SUBGROUP_HELPER_HEADER);
+  full_source.append(source);
+
+  std::optional<SPIRV::CodeVector> code;
+  switch (stage)
+  {
+  case ShaderStage::Vertex:
+    code = SPIRV::CompileVertexShader(full_source, APIType::Metal, glslang::EShTargetSpv_1_3);
+    break;
+  case ShaderStage::Geometry:
+    PanicAlertFmt("Tried to compile geometry shader for Metal, but Metal doesn't support them!");
+    break;
+  case ShaderStage::Pixel:
+    code = SPIRV::CompileFragmentShader(full_source, APIType::Metal, glslang::EShTargetSpv_1_3);
+    break;
+  case ShaderStage::Compute:
+    code = SPIRV::CompileComputeShader(full_source, APIType::Metal, glslang::EShTargetSpv_1_3);
+    break;
+  }
+  if (!code)
+    return "";
+
+  // clang-format off
+  static const spirv_cross::MSLResourceBinding resource_bindings[] = {
+      MakeResourceBinding(spv::ExecutionModelVertex,    0, 0, 1, 0, 0), // vs/ubo
+      MakeResourceBinding(spv::ExecutionModelVertex,    0, 1, 1, 0, 0), // vs/ubo
+      MakeResourceBinding(spv::ExecutionModelFragment,  0, 0, 0, 0, 0), // vs/ubo
+      MakeResourceBinding(spv::ExecutionModelFragment,  0, 1, 1, 0, 0), // vs/ubo
+      MakeResourceBinding(spv::ExecutionModelFragment,  1, 0, 0, 0, 0), // ps/samp0
+      MakeResourceBinding(spv::ExecutionModelFragment,  1, 1, 0, 1, 1), // ps/samp1
+      MakeResourceBinding(spv::ExecutionModelFragment,  1, 2, 0, 2, 2), // ps/samp2
+      MakeResourceBinding(spv::ExecutionModelFragment,  1, 3, 0, 3, 3), // ps/samp3
+      MakeResourceBinding(spv::ExecutionModelFragment,  1, 4, 0, 4, 4), // ps/samp4
+      MakeResourceBinding(spv::ExecutionModelFragment,  1, 5, 0, 5, 5), // ps/samp5
+      MakeResourceBinding(spv::ExecutionModelFragment,  1, 6, 0, 6, 6), // ps/samp6
+      MakeResourceBinding(spv::ExecutionModelFragment,  1, 7, 0, 7, 7), // ps/samp7
+      MakeResourceBinding(spv::ExecutionModelFragment,  1, 8, 0, 8, 8), // ps/samp8
+      MakeResourceBinding(spv::ExecutionModelFragment,  2, 0, 2, 0, 0), // ps/ssbo
+      MakeResourceBinding(spv::ExecutionModelGLCompute, 0, 1, 0, 0, 0), // cs/ubo
+      MakeResourceBinding(spv::ExecutionModelGLCompute, 1, 0, 0, 0, 0), // cs/output_image
+      MakeResourceBinding(spv::ExecutionModelGLCompute, 2, 0, 2, 0, 0), // cs/ssbo
+      MakeResourceBinding(spv::ExecutionModelGLCompute, 2, 1, 3, 0, 0), // cs/ssbo
+  };
+
+  spirv_cross::CompilerMSL::Options options;
+#if defined(TARGET_OS_MAC)
+  options.platform = spirv_cross::CompilerMSL::Options::macOS;
+#elif defined(TARGET_OS_IPHONE)
```

Comment:
Like the other target macros, this is always defined, so you probably want `#elif TARGET_OS_IPHONE`.